src/test/run-pass/existential_type.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(existential_type)]
+
+fn main() {
+    assert_eq!(foo().to_string(), "foo");
+    assert_eq!(bar1().to_string(), "bar1");
+    assert_eq!(bar2().to_string(), "bar2");
+    let mut x = bar1();
+    x = bar2();
+    assert_eq!(boo::boo().to_string(), "boo");
+    assert_eq!(cmp(), 5u32);
+    assert_eq!(cmp2(6u8), 6u8);
+}
+
+// single definition
+existential type Foo: std::fmt::Display;
+
+fn foo() -> Foo {
+    "foo"
+}
+
+// two definitions
+existential type Bar: std::fmt::Display;
+
+fn bar1() -> Bar {
+    "bar1"
+}
+
+fn bar2() -> Bar {
+    "bar2"
+}
+
+// definition in submodule
+existential type Boo: std::fmt::Display;
+
+mod boo {
+    pub fn boo() -> super::Boo {
+        "boo"
+    }
+}
+
+// generics
+existential type Cmp<T>: std::cmp::PartialEq<T> + std::fmt::Debug;
+
+fn cmp() -> Cmp<u32> {
```

Comment:
This should be an error according to the RFC. In order for type parameters to be applied correctly (and fully generally) there needs to be at least one definition site that leaves the parameters fully generic.

e.g.
```rust
existential type Cmp<A, B>;
struct Bar<A, B>(A, B);
fn foo() -> Cmp<u32, u32> {
    Bar(5u32, 6u32)
}
```
What is the definition of `Cmp`? It might seem like it is `type Cmp<A, B> = Bar<A, B>;`, but `type Cmp<A, B> = Cmp<B, A>;` might also be a valid choice.

Reply:
@oli-obk Yes, although it still has to be unified with whatever the "fully defining" usage is to make sure that it *can* work.